### ISSUE TITLE
Inlay hints multiply when using VSCode action "Rust Analyzer: Restart server"

### ISSUE THREAD

**Issue by Heinenen** [rust-lang/rust-analyzer#12836]
While writing some rust, I restarted the language server with the "Rust Analyzer: Restart server" action and encountered a bug.
My "workflow" to reproduce: 
1. Restart r-a server
2. Comment out a line
3. Save file
4. Uncomment same line

The bug is reproducible for me 100% of the time, even when I wait longer after restarting the language server.
Restarting the server multiple times increases the number of inlay hints multiple times (also seen in gif).

![inlay_hints2](https://user-images.githubusercontent.com/37484430/180051603-f1110dfd-1a3e-45cf-9af8-aafdb4f35161.gif)

**rust-analyzer version**: 0.4.1132-standalone (567a5e9ef 2022-07-18)

**rustc version**: rustc 1.62.0 (a8314ef7d 2022-06-27)

PS: When using the "Developer: Reload window" action, the inlay hints don't show up at all unless I change something in the file (with or without saving it). If you think that is worth another issue, I can file one for that as well.

**Comment by leighmcculloch** [comment]
I am experiencing the same thing with auto-completion and other things. Duplicate items start being displayed.

**Comment by alexschrod** [comment]
I'll chime in and say that I'm also experiencing this. It's like "restart" behaves more like "start another."

**Comment by Veykril** [comment]
odd, I can't reproduce this at all, not the duplication of inlay hints, nor inlay hints not appearing until i change the file ...

**Comment by tklajnscek** [comment]
+1 also had this happen to me recently, but only once and I couldn't make it happen again so I didn't report yet

**Comment by Heinenen** [comment]
I am also experiencing the same issues as @leighmcculloch, I just didn't notice it.
Another thing that appears multiple times is the outline and "Go to Symbol in Editor (`Ctrl + Shift + o`)"
If it helps, I am using Manjaro KDE with the open source build of VSCode.

**Comment by lnicola** [comment]
Just to make sure, how did you install the extension?

**Comment by leighmcculloch** [comment]
I didn't install rust-analyzer manually. It was installed via the vscode extension.

**Comment by lnicola** [comment]
How did you install the extension?

@Heinenen seems to have a recent build, at least of the server binary.

**Comment by jonathanpallant** [comment]
I see this on macOS for Arm and VS Code 1.69.2.

**Comment by Veykril** [comment]
This is very odd, I'm on VSCode 1.69.2, windows and cannot reproduce it with the latest server. When executing the Restart Server command do the inlay hints stay and a new set appears once restarted, or do the hints disappear and then re-appear duplicated?

**Comment by lnicola** [comment]
> although I ssh into an x86 Debian machine

Note that RA runs on the remote side. You should check the version number(s), just in case.

**Comment by leighmcculloch** [comment]
I am on VSCode 1.69.2, but I've seen this behavior on previous versions too.

I switch back and forth between the release and pre-release versions of rust-analyzer, and I'm not 100% confident, but I think I'm only seeing this on the pre-release version, v0.4.1136. I don't see this behavior on the latest release version, v.0.3.1131.

**Comment by Heinenen** [comment]
I installed my extension by downloading the .vsix from the [marketplace](https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer&ssr=false#version-history) and then installing it inside the Extensions tab inside vscode.
I have VSCode 1.69.2 aswell (for linux)

**Comment by Heinenen** [comment]
And I just disabled all the other extension and the bug persists.

**Comment by tklajnscek** [comment]
I am also on arm64 Mac, VSCode 1.69.2, extension v0.3.1131 installed from the marketplace

**Comment by Heinenen** [comment]
I installed rust-analyzer on my windows machine and both bugs appear there as well.

> When executing the Restart Server command do the inlay hints stay and a new set appears once restarted, or do the hints disappear and then re-appear duplicated?

One layer of inlay hints disappears and re-appears duplicated (so when restarting for the first time all inlay hints disappear and reappear duplicated). When there are multiple layers of inlay hints, only the "last" layer disappears and  is duplicated, the other layers stay. I think the gif I posted shows this behavior too.

**Comment by Veykril** [comment]
Okay, so I can't reliable produce it but I do get it sometimes, from the looks of it all requests are being duplicated by the client

**Comment by Veykril** [comment]
Okay, so restarting the server while its currently loading the workspace causes this, managed to duplicate it multiple times now
![image](https://user-images.githubusercontent.com/3757771/180567564-082a3c53-e8b6-4c93-86ad-3b7b1d1168ec.png)

**Comment by Veykril** [comment]
so, restarting the server into changing a file (and saving it, this is the important part) revives the old deactivated client, and the vscode changelogs reveal why
```
if an extension sends a notification or request before the client got started the client will auto start.
```

Now that was a pain to debug
